core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -3887,6 +3902,85 @@ public static ExpressionAnalyzer createWithoutSubqueries(
+            }
+        }
+        catch (RuntimeException e) {
+            throw semanticException(INVALID_DEFAULT_COLUMN_VALUE, literal, e, "'%s' is not a valid %s literal", literal, columnType.getDisplayName().toUpperCase(ENGLISH));
+        }
+    }
+
+    private static void checkDefaultColumnValue(Session session, PlannerContext plannerContext, Object value, Type type, Type literalType)
+    {
+        if (type instanceof BooleanType) {
+            checkArgument(value instanceof Boolean, "Value must be true or false '%s'".formatted(value));
+        }
+        if (type instanceof CharType charType) {
+            int targetLength = charType.getLength();
+            Slice slice = (Slice) value;
+            long actualLength = countCodePoints(slice);
+            checkArgument(targetLength >= actualLength, "Cannot truncate characters when casting value '%s' to %s".formatted(slice.toStringUtf8(), type));
+        }
+        if (type instanceof VarcharType varcharType) {
+            if (varcharType.isUnbounded()) {
+                return;
+            }
+            int targetLength = varcharType.getBoundedLength();
+            Slice slice = (Slice) value;
+            long actualLength = countCodePoints(slice);
+            checkArgument(targetLength >= actualLength, "Cannot truncate characters when casting value '%s' to %s".formatted(slice.toStringUtf8(), type));
+        }
+        if (type instanceof TimestampType timestampType) {
+            if (timestampType.isShort()) {
+                Long epochMicros = (Long) value;
+                int microOfSecond = floorMod(epochMicros, MICROSECONDS_PER_SECOND);
+                checkArgument(microOfSecond % Math.pow(10, TIMESTAMP_MICROS.getPrecision() - timestampType.getPrecision()) == 0, "Value too large");
+            }
+            else {
+                LongTimestamp longTimestamp = (LongTimestamp) value;
+                int picosOfMicro = longTimestamp.getPicosOfMicro();
+                checkArgument(picosOfMicro % Math.pow(10, TIMESTAMP_PICOS.getPrecision() - timestampType.getPrecision()) == 0, "Value too large");
+            }
+        }
+        if (type instanceof TimestampWithTimeZoneType timestampWithTimeZoneType) {
+            int precision = timestampWithTimeZoneType.getPrecision();
+            if (timestampWithTimeZoneType.isShort()) {
+                long epochMillis = unpackMillisUtc((long) value);
+                checkArgument(epochMillis % Math.pow(10, TIMESTAMP_TZ_MILLIS.getPrecision() - precision) == 0, "Value too large");
+            }
+            else {
+                LongTimestampWithTimeZone timestamp = (LongTimestampWithTimeZone) value;
+                int picosOfMilli = timestamp.getPicosOfMilli();
+                checkArgument(picosOfMilli % Math.pow(10, TIMESTAMP_TZ_PICOS.getPrecision() - precision) == 0, "Value too large");
+            }
+        }
```

Comment:
This logic will throw exceptions when a short timestamp literal is specified on long timestamp columns. Let me fix the implementation.